### PR TITLE
Apr 1204 extend notrequired condition to use contains

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionHandler.cs [hunk 1]
```diff
@@ -49,9 +52,13 @@ private static void RemoveInactivePages(ApplicationSection section)
             section.QnAData.Pages.RemoveAll(p => !p.Active);
         }
 
-        private static void RemovePagesBasedOnNotRequiredConditions(ApplicationSection section, JObject applicationData)
+        private  void RemovePagesBasedOnNotRequiredConditions(ApplicationSection section, JObject applicationData)
         {
-            section.QnAData.Pages.RemoveAll(p => p.NotRequiredConditions != null && p.NotRequiredConditions.Any(nrc => nrc.IsOneOf.Contains(applicationData[nrc.Field]?.Value<string>())));
+            // MFCMFC
```
Comment (David-B-Read) [inline src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionHandler.cs]: Can this be removed?

File: src/SFA.DAS.QnA.Application/Commands/SetPageAnswers/SetAnswersBase.cs [hunk 2]
```diff
@@ -138,27 +140,11 @@ protected Next GetNextActionForPage(Guid sectionId, string pageId)
                         if (!String.IsNullOrWhiteSpace(condition.QuestionTag))
                         {
                             var questionTagValue = applicationData[condition.QuestionTag];
+                            var questionTag = questionTagValue.Value<string>();
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Application/Commands/SetPageAnswers/SetAnswersBase.cs]: questionTagValue can be null. Please could you reinstate the null check

File: src/SFA.DAS.QnA.Application/Commands/SetPageAnswers/SetAnswersBase.cs [hunk 3]
```diff
@@ -215,13 +201,39 @@ protected Next GetNextActionForPage(Guid sectionId, string pageId)
             throw new ApplicationException($"Page {page.PageId}, in Sequence {page.SequenceId}, Section {page.SectionId} is missing a matching 'Next' instruction for Application {section.ApplicationId}");
         }
 
+        private static bool CheckAllConditionsSatisfied(Condition condition, string questionTag)
+        {
+            var allConditionsSatisfied = true;
+            
+
+            if (String.IsNullOrEmpty(questionTag))
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Application/Commands/SetPageAnswers/SetAnswersBase.cs]: Use string, not String  (it works differently at compiler level)

File: src/SFA.DAS.QnA.Application/Commands/SetPageAnswers/SetAnswersBase.cs [hunk 4]
```diff
@@ -215,13 +201,39 @@ protected Next GetNextActionForPage(Guid sectionId, string pageId)
...
+        }
+
         public Next FindNextRequiredAction(ApplicationSection section, Next nextAction, JObject applicationData)
         {
             if (section?.QnAData is null || nextAction is null || nextAction.Action != "NextPage") return nextAction;
 
             var isRequiredNextAction = true;
 
-            // Check here for any NotRequiredConditions on the next page.
+            // CheckAllConditionsSatisfied here for any NotRequiredConditions on the next page.
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Application/Commands/SetPageAnswers/SetAnswersBase.cs]: Typo in comment?

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionBySectionNoHandler.cs [hunk 5]
```diff
@@ -38,7 +41,7 @@ public async Task<HandlerResponse<Section>> Handle(GetSectionBySectionNoRequest
             return new HandlerResponse<Section>(_mapper.Map<Section>(section));
         }
 
-        private static void RemovePages(Data.Entities.Application application, ApplicationSection section)
+        private  void RemovePages(Data.Entities.Application application, ApplicationSection section)
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionBySectionNoHandler.cs]: Why is this not static any more?

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionBySectionNoHandler.cs [hunk 6]
```diff
@@ -51,9 +54,10 @@ private static void RemoveInactivePages(ApplicationSection section)
             section.QnAData.Pages.RemoveAll(p => !p.Active);
         }
 
-        private static void RemovePagesBasedOnNotRequiredConditions(ApplicationSection section, JObject applicationData)
+        private  void RemovePagesBasedOnNotRequiredConditions(ApplicationSection section, JObject applicationData)
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionBySectionNoHandler.cs]: Why is this not static any more?

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionHandler.cs [hunk 7]
```diff
@@ -36,7 +39,7 @@ public async Task<HandlerResponse<Section>> Handle(GetSectionRequest request, Ca
             return new HandlerResponse<Section>(_mapper.Map<Section>(section));
         }
 
-        private static void RemovePages(Data.Entities.Application application, ApplicationSection section)
+        private  void RemovePages(Data.Entities.Application application, ApplicationSection section)
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionHandler.cs]: Why is this not static anymore?

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionHandler.cs [hunk 8]
```diff
@@ -49,9 +52,10 @@ private static void RemoveInactivePages(ApplicationSection section)
             section.QnAData.Pages.RemoveAll(p => !p.Active);
         }
 
-        private static void RemovePagesBasedOnNotRequiredConditions(ApplicationSection section, JObject applicationData)
+        private  void RemovePagesBasedOnNotRequiredConditions(ApplicationSection section, JObject applicationData)
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionHandler.cs]: Why is this not static anymore?

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSequenceSections/GetSequenceSectionsHandler.cs [hunk 9]
```diff
@@ -40,7 +43,7 @@ public async Task<HandlerResponse<List<Section>>> Handle(GetSequenceSectionsRequ
             return new HandlerResponse<List<Section>>(sections);
         }
         
-        private static void RemovePages(Data.Entities.Application application, Section section)
+        private  void RemovePages(Data.Entities.Application application, Section section)
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Application/Queries/Sections/GetSequenceSections/GetSequenceSectionsHandler.cs]: Why is this not static anymore?

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSequenceSections/GetSequenceSectionsHandler.cs [hunk 10]
```diff
@@ -53,9 +56,11 @@ private static void RemoveInactivePages(Section section)
             section.QnAData.Pages.RemoveAll(p => !p.Active);
         }
 
-        private static void RemovePagesBasedOnNotRequiredConditions(Section section, JObject applicationData)
+        private  void RemovePagesBasedOnNotRequiredConditions(Section section, JObject applicationData)
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Application/Queries/Sections/GetSequenceSections/GetSequenceSectionsHandler.cs]: Why is this not static anymore?

File: src/SFA.DAS.QnA.Application/Services/INotRequiredProcessor.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿using System.Collections.Generic;
+using Newtonsoft.Json.Linq;
+using SFA.DAS.QnA.Api.Types.Page;
+
+namespace SFA.DAS.QnA.Application.Services
+{
+    public interface INotRequiredProcessor
+    {
+        bool NotRequired(List<NotRequiredCondition> notRequiredConditions, JObject applicationData);
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Application/Services/INotRequiredProcessor.cs]: Is List correct, should it be IEnumerable

File: src/SFA.DAS.QnA.Application/Services/INotRequiredProcessor.cs [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+        List<Page> PagesWithoutNotRequired(List<Page> pages, JObject applicationData);
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Application/Services/INotRequiredProcessor.cs]: Is List correct, should it be IEnumerable

File: src/SFA.DAS.QnA.Api/Startup.cs [hunk 13]
```diff
@@ -93,8 +94,8 @@ public void ConfigureServices(IServiceCollection services)
             services.AddTransient<IApplicationDataValidator, ApplicationDataValidator>();
             
             services.AddTransient<IEncryptionService, EncryptionService>();
+            services.AddTransient<INotRequiredProcessor, NotRequiredProcessor>();
```
Comment (martin-hornsby) [inline src/SFA.DAS.QnA.Api/Startup.cs]: Is Transient correct? Could it be made into a Singleton?

Comment (martin-hornsby) [reply src/SFA.DAS.QnA.Api/Startup.cs]: This is likely to be a mute point as all of the others are Transient 🤷‍♂

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionBySectionNoHandler.cs (hunk 5 above)
Comment (MarkFCain) [reply src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionBySectionNoHandler.cs]: cos a servie is injected.  I did play with noncreting the injected service at the top, but it goes through 2 layers of methods to be used, and is smellier than just not using a static

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionBySectionNoHandler.cs (hunk 6 above)
Comment (MarkFCain) [reply src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionBySectionNoHandler.cs]: see previous answer

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionHandler.cs (hunk 7 above)
Comment (MarkFCain) [reply src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionHandler.cs]: see previous answer

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionHandler.cs (hunk 8 above)
Comment (MarkFCain) [reply src/SFA.DAS.QnA.Application/Queries/Sections/GetSection/GetSectionHandler.cs]: see previous answer

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSequenceSections/GetSequenceSectionsHandler.cs (hunk 9 above)
Comment (MarkFCain) [reply src/SFA.DAS.QnA.Application/Queries/Sections/GetSequenceSections/GetSequenceSectionsHandler.cs]: see previous answer

File: src/SFA.DAS.QnA.Application/Queries/Sections/GetSequenceSections/GetSequenceSectionsHandler.cs (hunk 10 above)
Comment (MarkFCain) [reply src/SFA.DAS.QnA.Application/Queries/Sections/GetSequenceSections/GetSequenceSectionsHandler.cs]: ditto

File: src/SFA.DAS.QnA.Api/Startup.cs (hunk 13 above)
Comment (MarkFCain) [reply src/SFA.DAS.QnA.Api/Startup.cs]: Still seems like a singleton to me tho 👍

**Pull request merged**